### ISSUE TITLE
barchart label not working in milestone 0.5

### ISSUE THREAD

**Issue by germanftorres** [DmitryBaranovskiy/g.raphael#129]
Hi!

I'm trying to label a single series barchart and it simply doesn't work. I have modified the test/barchart.html example trying to label the first chart. It gives an exception an the other three charts don't get rendered. It appears as the labelise method gets called on an object which doesn't support it. I have tried to fix de issue but I'm not really good at javascript.

Thanks

Germán

```
        window.onload = function () {
            var r = Raphael("holder"),
                data1 = [[55, 20, 13, 32, 5, 1, 2, 10], [10, 2, 1, 5, 32, 13, 20, 55], [12, 20, 30]],
                data2 = [[55, 20, 13, 32, 5, 1, 2, 10], [10, 2, 1, 5, 32, 13, 20, 55], [12, 20, 30]],
                data3 = [[55, 20, 13, 32, 5, 1, 2, 10], [10, 2, 1, 5, 32, 13, 20, 55], [12, 20, 30]],
                txtattr = { font: "12px 'Fontin Sans', Fontin-Sans, sans-serif" };

            r.text(160, 10, "Single Series Chart").attr(txtattr);
            r.text(480, 10, "Multiline Series Chart").attr(txtattr);
            r.text(160, 250, "Multiple Series Stacked Chart").attr(txtattr);
            r.text(480, 250, 'Multiline Series Stacked Vertical Chart. Type "round"').attr(txtattr);

            var chart = r.barchart(10, 10, 300, 220, [[55, 20, 13, 32, 5, 1, 2, 10]], 0, {type: "sharp"});

            //
            // here we get the error 'object does not support labelise property or method'
            //
            chart.label(["Uno", "Dos", "Tres", "Cuatro", "Cinco", "Seis", "Siete", "Ocho"]);

            r.barchart(330, 10, 300, 220, data1);
            r.barchart(10, 250, 300, 220, data2, {stacked: true});
            r.barchart(330, 250, 300, 220, data3, {stacked: true, type: "round"});
        };
```

**Comment by rzurad** [comment]
Ah, looks like when we moved from a namespace to a prototype, we mucked up a few references. Looks like there's a few logical errors in how bar charts are trying to apply labels. I'll look into it.

**Comment by karolk** [comment]
Just to let you know my current stance on the labels - from what I noticed textual labels claim all space there is for the chart, so if they are really long they will squish the graph. I am planning to prepare specific test cases and deal with this when I am finished with docs.
I am also tempted to experiment with dividing the code into axis, legend and actual graph, because they are intertwined at the moment. Axis already lives on the prototype and legend code could be moved there.

**Comment by rzurad** [comment]
I agree. There is lots and lots of room to refactor the codebase into something much more modular, that way there is less coupled code that all the different graph types can leverage.

**Comment by amuraco** [comment]
Has there been any progress on fixing these issues? Currently the labels on a barchart just throw an error.

**Comment by nullvariable** [comment]
I'm still getting 

``` javascript
Object #<a> has no method 'labelise' 
```

when trying to add labels, is this going to be fixed or are we on our own with this?

**Comment by pierrenel** [comment]
Yep, still having this issue.

**Comment by torgeir** [comment]
+1

**Comment by theirishpenguin** [comment]
+1

**Comment by saimn** [comment]
I have fixed barchart labels for my own purpose, the commit is there:
saimn/c2c-stats@aa4b43224400c9162ca30a443b32284a4d604f90

I'm not a javascript expert and I didn't test all barchart options but it works for me : it works with stacked (my commit msg is wrong) and non-stacked, hbarchart and vbarchart. 
However there is another bug with tooltips and raphael 2.1.0. With raphael 2.0.2 it's working fine.

**Comment by theirishpenguin** [comment]
Thanks @saimn You're fix kind of worked in my case. When I applied it, the actual column values appeared on top of the bars (rather than a distinct legend that I could specific underneath them). What I'd like is a barchart like outlined here http://masonoise.wordpress.com/2009/12/29/bar-and-pie-charts-with-raphaeljs though I think the blog poster uses on older version of raphael and graphael (0.4).

I'm using the interactive barchart and like you said, I had to use raphael 2.0.2 to see any effect.

**Comment by saimn** [comment]
@theirishpenguin for `barchart`, labels are shown by default at the top of the bars. You can use `.label( [...], true)` to put labels at the bottom. 
For `vbarchart` there is a corresponding option to show lables on the right instead of left which is the default.

**Comment by theirishpenguin** [comment]
@saimn Thanks for clarifying that! It now works as expected, with one small change - I had to wrap the labels inside an extra pair of square brackets, for example

```
r.barchart(
  10, 10, 300, 220,
  [[55, 20, 13, 32, 5, 1, 2, 10, 11,44, 64, 23]],
  {type: "round"}
).hover(fin, fout).label([['J','F','M','A','M','J','J','A', 'S', 'O', 'N','D']], true);
```

Thanks again. Was banging my head against a wall until you provided your solution!

**Comment by abhisec** [comment]
Is this being maintained? The bar chart labels not working is kind of critical? Are the authors maintaining this?

**Comment by baileyspace** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by ghost** [comment]
Thanks @saimn, I've gotten the correct labels to show but they are straight after another rather than below their individual bars, and I still get an error

```
TypeError: 'undefined' is not an object (evaluating 'bars[j][i].x')
```

Help would be appreciated :)

**Comment by mattleff** [comment]
+1

**Comment by wittxiao** [comment]
Anybody know the solution? I got the same problem with 0.51

**Comment by mcasperson** [comment]
+1

**Comment by tian-yi** [comment]
+1

**Comment by tian-yi** [comment]
For people who's still looking for solutions, try using https://github.com/jhurt/g.raphael/blob/master/g.bar.js which fixes this issue and worked pretty well  for me

**Comment by mikeshultz** [comment]
+1

**Comment by trollfred** [comment]
+1 and solution from tianyi33 works

**Comment by dlaxar** [comment]
+1 this is still unresolved...

**Comment by flyon** [comment]
+1 yes still unresolved ... defenitly need labels, and I'm getting the same error "Uncaught TypeError: Object #<c> has no method 'labelise' "

```
        var chart = r.barchart(10, 10, 550, 300, [
            [5.5, 2.0, 7.6, 9.5]
        ]).hover(fin, fout).label([['a','b','c','d']]);
```

**Comment by jayarjo** [comment]
+1

**Comment by villian** [comment]
still problem with label s
